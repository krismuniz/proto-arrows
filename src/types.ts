export interface Point {
  x: number;
  y: number;
}

export interface Box {
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface CubicBezier {
  start: Point;
  control1: Point;
  control2: Point;
  end: Point;
}
