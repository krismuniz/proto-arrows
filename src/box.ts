import { Box, Point } from './types'

/**
 * Get a list of points set up at the center
 * of every side of the box (left, right, top, bottom)
 */
export function getBoxBounds (box: Box): [Point, Point, Point, Point] {
  return [
    { x: box.x, y: box.y + box.h / 2 }, // left
    { x: box.x + box.w, y: box.y + box.h / 2 }, // right
    { x: box.x + box.w / 2, y: box.y }, // top
    { x: box.x + box.w / 2, y: box.y + box.h } // bottom
  ]
}

/**
 * Find the ideal points to draw an edge
 * between the two boxes by finding the
 * shortest linear distance between the
 * two.
 */
export function getIdealBoxSides (
  startBox: Box,
  endBox: Box
): {
  startPoint: Point
  endPoint: Point
} {
  const startPts = getBoxBounds(startBox)
  const endPts = getBoxBounds(endBox)

  let minDistanceSource: [number, Point] = [Infinity, { x: 0, y: 0 }]
  let minDistanceTarget: [number, Point] = [Infinity, { x: 0, y: 0 }]

  for (const pointA of startPts) {
    for (const pointB of endPts) {
      const distance = Math.hypot(pointA.x - pointB.x, pointA.y - pointB.y)

      if (distance < minDistanceSource[0]) {
        minDistanceSource = [distance, pointA]
      }

      if (distance < minDistanceTarget[0]) {
        minDistanceTarget = [distance, pointB]
      }
    }
  }

  return {
    startPoint: minDistanceSource[1],
    endPoint: minDistanceTarget[1]
  }
}

/**
 * Infer if a point is on the left or right side of a box
 */
export function isPointOnLeftOrRightSide (point: Point, box: Box): boolean {
  return point.x === box.x || point.x === box.x + box.w
}
