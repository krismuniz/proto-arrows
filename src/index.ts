export type { Box, Point, CubicBezier } from './types'
export { getBoxToBoxCurve, getCurve, getCubicBezierSVGPath } from './path'
export { interpolateCubicBezier, interpolateCubicBezierAngle } from './interpolation'
