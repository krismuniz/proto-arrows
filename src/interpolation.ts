import { CubicBezier, Point } from './types'

/**
 * Given the definition of a cubic bezier: a start point, two control points,
 * and end point, get the point at a given time `t` (where `0 <= t <= 1`).
 *
 * For example, at t = 0, this function returns the point at the start of the
 * curve, at t = 0.5, it returns the point midway through the curve and at
 * t = 1 it returns the point at the end of the curve.
 *
 * B(t) = (1 - t)^3P0 + 3(1 - t)^2tP1 + 3(1 - t)t^2P2 + t^3P3
 * Adapted from https://github.com/pbeshai/vis-utils
 */
export function interpolateCubicBezier ({
  start,
  control1,
  control2,
  end
}: CubicBezier, t: number): Point {
  /**
   * Get the point on the curve at a given t,
   * where t is a number between 0 and 1.
   *
   * 0 is the start point, 1 is the end point.
   */
  return {
    x: Math.pow(1 - t, 3) * start.x +
      3 * Math.pow(1 - t, 2) * t * control1.x +
      3 * (1 - t) * Math.pow(t, 2) * control2.x +
      Math.pow(t, 3) * end.x,
    y: Math.pow(1 - t, 3) * start.y +
      3 * Math.pow(1 - t, 2) * t * control1.y +
      3 * (1 - t) * Math.pow(t, 2) * control2.y +
      Math.pow(t, 3) * end.y
  }
}

/**
 * Given the definition of a cubic bezier: a start point, two control points,
 * and end point, get the angle at a given time `t` (where `0 <= t <= 1`).
 *
 * For example, at t = 0, this function returns the angle at the start
 * point, at t = 0.5, it returns the angle midway through the curve and at
 * t = 1 it returns the angle at the end of the curve (useful for things like
 * arrowheads). The angles are in degrees.
 *
 * B'(t) = 3(1- t)^2(P1 - P0) + 6(1 - t)t(P2 - P1) + 3t^2(P3 - P2)
 *
 * Adapted from https://github.com/pbeshai/vis-utils
 */
export function interpolateCubicBezierAngle ({
  start,
  control1,
  control2,
  end
}: CubicBezier, t: number) {
  /**
   * Get the angle of the point on the curve at a given t,
   * where t is a number between 0 and 1.
   *
   * 0 is the start point, 1 is the end point.
   */
  const tangentX = 3 * Math.pow(1 - t, 2) * (control1.x - start.x) +
      6 * (1 - t) * t * (control2.x - control1.x) +
      3 * Math.pow(t, 2) * (end.x - control2.x)
  const tangentY = 3 * Math.pow(1 - t, 2) * (control1.y - start.y) +
      6 * (1 - t) * t * (control2.y - control1.y) +
      3 * Math.pow(t, 2) * (end.y - control2.y)

  return Math.atan2(tangentY, tangentX) * (180 / Math.PI)
}
