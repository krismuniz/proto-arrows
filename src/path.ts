import { Box, CubicBezier, Point } from './types'
import { getIdealBoxSides, isPointOnLeftOrRightSide } from './box'

/**
 * Find the center of two points using rudimentary
 * linear interpolation.
 */
export function getLineCenter (pointA: Point, pointB: Point): Point {
  return {
    x: (pointA.x + pointB.x) / 2,
    y: (pointA.y + pointB.y) / 2
  }
}

/**
 * Given a cubic bezier curve, produce its corresponding
 * SVG path string.
 */
export function getCubicBezierSVGPath (bezier: CubicBezier): string {
  const { start, control1, control2, end } = bezier

  return [
    `M${start.x},${start.y}`,
    `C${control1.x},${control1.y} ${control2.x},${control2.y}`,
    `${end.x},${end.y}`
  ].join(' ')
}

/**
 * Given two points, produce a cubic bezier curve that
 * links them.
 */
export function getCurve (start: Point, end: Point, options?: { flip?: boolean }): CubicBezier {
  const center = getLineCenter(start, end)
  const controlPoints: [Point, Point] =
    options?.flip
      ? [
          {
            x: center.x,
            y: start.y
          },
          {
            x: center.x,
            y: end.y
          }
        ]
      : [
          {
            x: start.x,
            y: center.y
          },
          {
            x: end.x,
            y: center.y
          }
        ]

  return { start: start, control1: controlPoints[0], control2: controlPoints[1], end: end }
}

export function getBoxToBoxCurve (startBox: Box, endBox: Box): CubicBezier {
  const { startPoint, endPoint } = getIdealBoxSides(startBox, endBox)

  return getCurve(
    startPoint,
    endPoint,
    {
      // Flip the curve if the `startPoint` is on the left/right
      // side of the `startBox` AND the `endPoint` is on the
      // left/right side of the `endBox`.
      //
      // In the future we'll make this an option.
      flip: isPointOnLeftOrRightSide(startPoint, startBox) &&
      isPointOnLeftOrRightSide(endPoint, endBox)
    }
  )
}
