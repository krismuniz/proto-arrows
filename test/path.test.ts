import { getBoxToBoxCurve, getCubicBezierSVGPath, getCurve, getLineCenter } from '../src/path'
import { easeInOutCubic } from './beziers'

describe('getLineCenter', function () {
  it('properly gets the center point using linear interpolation', function () {
    expect(getLineCenter({ x: 0, y: 0 }, { x: 10, y: 10 })).toEqual({ x: 5, y: 5 })
    expect(getLineCenter({ x: 0, y: 0 }, { x: -10, y: -10 })).toEqual({ x: -5, y: -5 })
    expect(getLineCenter({ x: 0, y: 0 }, { x: 0, y: 10 })).toEqual({ x: 0, y: 5 })
    expect(getLineCenter({ x: 0, y: 0 }, { x: 10, y: 0 })).toEqual({ x: 5, y: 0 })
  })
})

describe('getCubicBezierSVGPath', function () {
  it('returns a proper svg path in the correct format', function () {
    expect(getCubicBezierSVGPath(easeInOutCubic)).toEqual('M0,100 C65,0 35,1 100,100')
  })
})

describe('getCurve', function () {
  it('returns the right curve given two points', function () {
    expect(getCurve({ x: 0, y: 0 }, { x: 10, y: 10 })).toMatchSnapshot()
    expect(getCurve({ x: 0, y: 0 }, { x: -10, y: -10 })).toMatchSnapshot()
    expect(getCurve({ x: 0, y: 0 }, { x: 0, y: 10 })).toMatchSnapshot()
    expect(getCurve({ x: 0, y: 0 }, { x: 10, y: 0 })).toMatchSnapshot()
  })
})

describe('getBoxToBoxCurve', function () {
  it('returns the right curve given two boxes', function () {
    expect(getBoxToBoxCurve({ x: 0, y: 0, w: 4, h: 4 }, { x: 0, y: 0, w: 4, h: 4 })).toMatchSnapshot()
    // on the right side of the other box
    expect(getBoxToBoxCurve({ x: 0, y: 0, w: 4, h: 4 }, { x: 4, y: 0, w: 4, h: 4 })).toMatchSnapshot()
    // above the other one
    expect(getBoxToBoxCurve({ x: 0, y: 0, w: 4, h: 4 }, { x: 0, y: -4, w: 4, h: 4 })).toMatchSnapshot()
    // below the other one
    expect(getBoxToBoxCurve({ x: 0, y: 0, w: 4, h: 4 }, { x: 0, y: 4, w: 4, h: 4 })).toMatchSnapshot()
    // left side of the other box
    expect(getBoxToBoxCurve({ x: 0, y: 0, w: 4, h: 4 }, { x: -4, y: 0, w: 4, h: 4 })).toMatchSnapshot()
  })
})
