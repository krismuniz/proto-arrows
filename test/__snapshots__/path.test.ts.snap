// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBoxToBoxCurve returns the right curve given two boxes 1`] = `
Object {
  "control1": Object {
    "x": 0,
    "y": 2,
  },
  "control2": Object {
    "x": 0,
    "y": 2,
  },
  "end": Object {
    "x": 0,
    "y": 2,
  },
  "start": Object {
    "x": 0,
    "y": 2,
  },
}
`;

exports[`getBoxToBoxCurve returns the right curve given two boxes 2`] = `
Object {
  "control1": Object {
    "x": 4,
    "y": 2,
  },
  "control2": Object {
    "x": 4,
    "y": 2,
  },
  "end": Object {
    "x": 4,
    "y": 2,
  },
  "start": Object {
    "x": 4,
    "y": 2,
  },
}
`;

exports[`getBoxToBoxCurve returns the right curve given two boxes 3`] = `
Object {
  "control1": Object {
    "x": 2,
    "y": 0,
  },
  "control2": Object {
    "x": 2,
    "y": 0,
  },
  "end": Object {
    "x": 2,
    "y": 0,
  },
  "start": Object {
    "x": 2,
    "y": 0,
  },
}
`;

exports[`getBoxToBoxCurve returns the right curve given two boxes 4`] = `
Object {
  "control1": Object {
    "x": 2,
    "y": 4,
  },
  "control2": Object {
    "x": 2,
    "y": 4,
  },
  "end": Object {
    "x": 2,
    "y": 4,
  },
  "start": Object {
    "x": 2,
    "y": 4,
  },
}
`;

exports[`getBoxToBoxCurve returns the right curve given two boxes 5`] = `
Object {
  "control1": Object {
    "x": 0,
    "y": 2,
  },
  "control2": Object {
    "x": 0,
    "y": 2,
  },
  "end": Object {
    "x": 0,
    "y": 2,
  },
  "start": Object {
    "x": 0,
    "y": 2,
  },
}
`;

exports[`getCurve returns the right curve given two points 1`] = `
Object {
  "control1": Object {
    "x": 0,
    "y": 5,
  },
  "control2": Object {
    "x": 10,
    "y": 5,
  },
  "end": Object {
    "x": 10,
    "y": 10,
  },
  "start": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`getCurve returns the right curve given two points 2`] = `
Object {
  "control1": Object {
    "x": 0,
    "y": -5,
  },
  "control2": Object {
    "x": -10,
    "y": -5,
  },
  "end": Object {
    "x": -10,
    "y": -10,
  },
  "start": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`getCurve returns the right curve given two points 3`] = `
Object {
  "control1": Object {
    "x": 0,
    "y": 5,
  },
  "control2": Object {
    "x": 0,
    "y": 5,
  },
  "end": Object {
    "x": 0,
    "y": 10,
  },
  "start": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`getCurve returns the right curve given two points 4`] = `
Object {
  "control1": Object {
    "x": 0,
    "y": 0,
  },
  "control2": Object {
    "x": 10,
    "y": 0,
  },
  "end": Object {
    "x": 10,
    "y": 0,
  },
  "start": Object {
    "x": 0,
    "y": 0,
  },
}
`;
